sdk/keyvault/azure-keyvault-administration/azure/keyvault/administration/_internal/async_challenge_auth_policy.py

```diff
@@ -37,12 +68,87 @@ class AsyncChallengeAuthPolicy(AsyncBearerTokenCredentialPolicy):
     """
 
     def __init__(self, credential: AsyncTokenCredential, *scopes: str, **kwargs: Any) -> None:
-        super().__init__(credential, *scopes, **kwargs)
+        # Pass `enable_cae` so `enable_cae=True` is always passed through self.authorize_request
+        super().__init__(credential, *scopes, enable_cae=True, **kwargs)
```

Comment:
Is changing the default value of enable_cae to True a breaking change?

Reply:
I wouldn't think so, since the feature hasn't been available for Key Vault in the past (and won't be available until after this version is released).

Reply:
You mean in this case, even client asks for a CAE token, service still returns a non-cae token?

Reply:
Yes. CP1 means "I can handle CAE challenges" not "give me a CAE token". The tenant decides what kind of token to issue and if the RP doesn't support CAE, the kind of token is irrelevant.

Reply:
Hmm... interesting, then there may be more issues.

We cache the token based on the input.

This means we may cache a non-cae token in cae cache and other non-cae credential will not be able to share the correct token with it.?

Reply:
We cache CAE tokens separately, see e.g. [here](https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/identity/azure-identity/azure/identity/_internal/msal_credentials.py#L73) and [here](https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/identity/azure-identity/azure/identity/_internal/msal_credentials.py#L98)

Reply:
How is that a problem?

Reply:
Since we don't have control over whether a service supports CAE, isn't that the best we can do to sort the tokens? i.e. one cache for tokens we attempted to enable CAE for, and another cache for tokens we didn't? We may end up with non-CAE tokens in the "CAE cache", but I don't see how that would be a problem for functionality.

Reply:
> How is that a problem?

As you pointed out above, multiple objects can share the cache.

Thinking about multi-version kv SDK running side by side. The old behavior they could share the cache but the new behavior would be they are not able to share. (one searches in non-cae cache while the other searches in cae cache).

Reply:
@xiangyan99 in that scenario there would be two tokens instead of one shared with the cache, so one client could make a token request that it wouldn't have made before. I don't think that's necessarily breaking though, and I don't see a problem with that cache scenario.

Reply:
Btw, to make sure we are on the same page, you know that if happened that some custom credentials raises when you added enable_cae in get_token, right?

Reply:
You mean that a custom credential implementation could raise an error when `enable_cae` is passed to `get_token`? Are you referring to https://github.com/Azure/azure-sdk-for-python/pull/31546?

Reply:
We made https://github.com/Azure/azure-sdk-for-python/pull/31546 because some custom credentials raises when you added enable_cae in get_token. We only pass enable_cae when required.

Reply:
When I said "required" I meant the case that the users explicitly specified enable_cae=True. We don't want to make it fail in the case that customers are even not aware cae is a thing.

Reply:
Got it. It's a bit frustrating since `**kwargs` have been documented in the `TokenCredential` protocol for at least two years, and `enable_cae` has been in the protocol for at least a year (including in our minimum requirement of `azure-core` for KV libraries), but I do see that it could potentially be breaking.

To make it non-breaking, would we need to add an `enable_cae` kwarg to our client somewhere, with a default value of False? I don't want to expose this to users through KV, but I don't see how we could otherwise make it completely non-breaking without doing something like a `try/except` on token requests.

Reply:
@xiangyan99 do you mean that applications should always be responsible for passing `enable_cae`? Even if a Key Vault client can handle CAE challenges, it shouldn't pass `enable_cae=True` to a credential because that credential may raise?

Reply:
Sorry, I'm still not following you. How would our Key Vault client passing `enable_cae=True` break applications?

Reply:
In your code, you will call get_token(enable_cae=True).

It happened that some custom credentials did not handle argument `enable_cae` (any new argument) correctly. They passed enable_cae into transport and the transport raised.

Reply:
That would be a bug in the custom credential, no? The TokenCredential protocol requires `get_token` to accept `**kwargs`

Reply:
Right. The root cause was the custom credential did not implement get_token decently enough. But what happened was after we passed in enable_cae, cli crashed.

That's the reason we need to take it carefully and want to do the best to avoid the case from happening again.

Reply:
That's a CLI bug though. `get_token` accepts `**kwargs`, so implementations are responsible for handling unexpected args gracefully.

Reply:
Agreed this is a bug on the custom credential side.

But the end user experience is their code worked before and after the update, the code stops working.

And the reason is we added a feature that they don't need.

Reply:
Is the CLI still unable to accept `enable_cae`? It does look like they still [pass `**kwargs` into MSAL methods](https://github.com/Azure/azure-cli/blob/0a35d5f7f9074bd39dce85d331712abb59d77afb/src/azure-cli-core/azure/cli/core/auth/msal_credentials.py#L60-L61) without accounting for the argument, in the same way that produced the original issue.

Reply:
Sure, that's a good idea in general.

In this particular case, Key Vault clients should pass `enable_cae=True` unconditionally because doing so isn't a breaking change and we want applications to be secure by default.

Reply:
Unfortunately, I do think the CLI could be preventing us from always providing `enable_cae`. Very frustrating, but I don't want to break a major SDK consumer even if the bug is on their side.

@chlowell I could either bring this up with the CLI team and hope they fix it soon -- which I don't think is likely, given that they didn't adjust to Core's prior change -- or we could make `enable_cae` a client constructor option that defaults to False. Would the latter be acceptable, at least for the time being, so that we can ship this support this month?

Reply:
I believe they already have this fixed + they have not migrated to use key vault track 2 in their code.

Let's give them a heads up before we ship the code.

Reply:
We've confirmed that the CLI's credentials are still incompatible with `enable_cae`. However, their KV dependencies are pinned, so they wouldn't be exposed to new releases with this change; they told us to go ahead with releases as planned. Resolving this thread.